pkg/ruler/rule_concurrency.go

```diff
@@ -0,0 +1,213 @@
+		maxGlobalConcurrency: maxGlobalConcurrency,
+		limits:               limits,
+		globalConcurrency:    *semaphore.NewWeighted(maxGlobalConcurrency),
+		tenantConcurrency:    make(map[string]*DynamicSemaphore),
+	}
+}
+
+// Done releases a slot from the concurrency controller.
+func (c *MultiTenantConcurrencyController) Done(ctx context.Context) {
+	c.globalConcurrency.Release(1)
+	c.metrics.CurrentConcurrency.Dec()
+	c.metrics.ReleaseTotal.Inc()
+
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error extracting org id from context while releasing a slot", "err", err)
+		return
+	}
+
+	c.mtx.Lock()
+	tc, ok := c.tenantConcurrency[userID]
+	c.mtx.Unlock()
+
+	if ok {
+		tc.Release()
+	} else {
+		c.logger.Log("msg", "tenant concurrency controller: tenant not found", "tenant", userID)
+	}
+}
+
+// Allow tries to acquire a slot from the concurrency controller.
+func (c *MultiTenantConcurrencyController) Allow(ctx context.Context, group *rules.Group, rule rules.Rule) bool {
+	c.metrics.AcquireTotal.Inc()
+	// To allow a rule to be executed concurrently, we need 3 conditions:
+	// 1. The rule group must be at risk of missing its evaluation.
+	// 2. The rule must not have any rules that depend on it.
+	// 3. The rule itself must not depend on any other rules.
+	if !isGroupAtRisk(group) {
+		return false
+	}
+
+	if !isRuleIndependent(rule) {
+		return false
+	}
+
+	// Next, try to acquire a global concurrency slot.
+	if !c.globalConcurrency.TryAcquire(1) {
+		c.metrics.AcquireFailedTotal.Inc()
+		return false
+	}
+	c.metrics.CurrentConcurrency.Inc()
+
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error extracting org id from context while acquiring a slot", "err", err)
+		c.globalConcurrency.Release(1)
+		c.metrics.CurrentConcurrency.Dec()
+		c.metrics.AcquireFailedTotal.Inc()
+		return false
+	}
+
+	c.mtx.Lock()
+	tc, ok := c.tenantConcurrency[userID]
+	if !ok {
+		tc = NewDynamicSemaphore(c.logger, func() int64 {
+			return c.limits.RulerMaxConcurrentRuleEvaluationsPerTenant(userID)
+		})
+		c.tenantConcurrency[userID] = tc
+	}
+	c.mtx.Unlock()
+
+	if tc.TryAcquire() {
+		return true
+	}
+
+	// If we can't acquire a tenant slot, release the global slot.
+	c.globalConcurrency.Release(1)
+	c.metrics.CurrentConcurrency.Dec()
+	c.metrics.AcquireFailedTotal.Inc()
+	return false
+}
+
+// isRuleIndependent checks if the rule is independent of other rules.
+func isRuleIndependent(rule rules.Rule) bool {
+	return rule.NoDependentRules() && rule.NoDependencyRules()
+}
+
+// isGroupAtRisk checks if the rule group's last evaluation time is within the risk threshold.
+func isGroupAtRisk(group *rules.Group) bool {
+	interval := group.Interval().Seconds()
+	lastEvaluation := group.GetEvaluationTime().Seconds()
+
+	return lastEvaluation < interval*IntervalToEvaluationThreshold/100
```

Comment:
Isn't the opposite? The group is at risk when the last evaluation took longer (>) than 50% of interval? Can we unit test this function too, please?

Reply:
I tried unit testing it, but the interval is a private property and there's no way to cleverly modifying the value. Let me give it another try.

Reply:
You were right, I must have made a silly mistake - gave it another try and tried to test it in https://github.com/grafana/mimir/pull/8146/commits/f419dc09cc88b80e4db9b583870e88dffdf32266

I also did a testing of this in prometheus to make sure I wasn't crazy.

```
diff --git a/rules/manager.go b/rules/manager.go
index 9e5b33fbc..655d791f2 100644
--- a/rules/manager.go
+++ b/rules/manager.go
@@ -37,6 +37,8 @@ import (
 	"github.com/prometheus/prometheus/util/strutil"
 )
 
+const IntervalToEvaluationThreshold = 50.00
+
 // QueryFunc processes PromQL queries.
 type QueryFunc func(ctx context.Context, q string, t time.Time) (promql.Vector, error)
 
@@ -477,22 +479,39 @@ func newRuleConcurrencyController(maxConcurrency int64) RuleConcurrencyControlle
 	}
 }
 
-func (c *concurrentRuleEvalController) Allow(_ context.Context, _ *Group, rule Rule) bool {
+func (c *concurrentRuleEvalController) Allow(_ context.Context, group *Group, rule Rule) bool {
 	// To allow a rule to be executed concurrently, we need 3 conditions:
 	// 1. The rule must not have any rules that depend on it.
 	// 2. The rule itself must not depend on any other rules.
 	// 3. If 1 & 2 are true, then and only then we should try to acquire the concurrency slot.
-	if rule.NoDependentRules() && rule.NoDependencyRules() {
-		return c.sema.TryAcquire(1)
+	if !isRuleIndependent(rule) {
+		return false
 	}
 
-	return false
+	if !isGroupAtRisk(group) {
+		return false
+	}
+
+	return c.sema.TryAcquire(1)
 }
 
 func (c *concurrentRuleEvalController) Done(_ context.Context) {
 	c.sema.Release(1)
 }
 
+// isRuleIndependent checks if the rule is independent of other rules.
+func isRuleIndependent(rule Rule) bool {
+	return rule.NoDependentRules() && rule.NoDependencyRules()
+}
+
+// isGroupAtRisk checks if the rule group's last evaluation time is within the risk threshold.
+func isGroupAtRisk(group *Group) bool {
+	interval := group.Interval().Seconds()
+	lastEvaluation := group.GetEvaluationTime().Seconds()
+
+	return lastEvaluation >= interval*IntervalToEvaluationThreshold/100
+}
+
 // sequentialRuleEvalController is a RuleConcurrencyController that runs every rule sequentially.
 type sequentialRuleEvalController struct{}
 
diff --git a/rules/manager_test.go b/rules/manager_test.go
index 51239e6c9..65a956a58 100644
--- a/rules/manager_test.go
+++ b/rules/manager_test.go
@@ -2154,3 +2154,37 @@ func optsFactory(storage storage.Storage, maxInflight, inflightQueries *atomic.I
 		},
 	}
 }
+
+func Test_concurrentRuleEvalController_GroupLastEvaluationOverThreshold(t *testing.T) {
+	tc := map[string]struct {
+		group    *Group
+		expected bool
+	}{
+		"group last evaluation greater than interval": {
+			group:    &Group{interval: 10 * time.Second, evaluationTime: 15 * time.Second},
+			expected: true,
+		},
+		"group last evaluation less than interval": {
+			group:    &Group{interval: 10 * time.Second, evaluationTime: 1 * time.Second},
+			expected: false,
+		},
+		"group last evaluation lower than concurrency trigger threshold": {
+			group:    &Group{interval: 10 * time.Second, evaluationTime: 4 * time.Second},
+			expected: false,
+		},
+		"group last evaluation higher than concurrency trigger threshold": {
+			group:    &Group{interval: 10 * time.Second, evaluationTime: 6 * time.Second},
+			expected: true,
+		},
+		"group last evaluation exactly at concurrency trigger threshold": {
+			group:    &Group{interval: 10 * time.Second, evaluationTime: 5 * time.Second},
+			expected: true,
+		},
+	}
+
+	for name, tt := range tc {
+		t.Run(name, func(t *testing.T) {
+			require.Equal(t, tt.expected, isGroupAtRisk(tt.group))
+		})
+	}
+}

```